### PR TITLE
Configurable feature seeds

### PR DESCRIPTION
**Pull request by kennytv**
Co-authored-by: Thonk <30448663+ExcessiveAmountsOfZombies@users.noreply.github.com>

This fixes issues, reduces diff by a lot, and adds default filling, so I felt like opening a few is easier than requesting a dozen changes and going on with the original pr.

No seeds set:
No seeds

disk_clay set:
disk_clay set

Randomize everything option (after deleting the manually added disk_clay)
Randomized

### CODE REVIEW

**Comment by ExcessiveAmountsOfZombies** [comment]
thanks, i had tried rebasing it one day, spent a few hours on it and it wouldn't fix the conflicts and just gave up. figured it'd probably be easier for someone who maintains the project to rebase it

**Comment by kennytv** [comment]
> thanks, i had tried rebasing it one day, spent a few hours on it and it wouldn't fix the conflicts and just gave up. figured it'd probably be easier for someone who maintains the project to rebase it

Good to know I have your blessing - thanks for the original pr and code, even if it's continued here now!

**Comment by MiniDigger** [comment]
please attach a docs PR 😈

File: patches/server/0802-Configurable-feature-seeds.patch [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ import net.minecraft.world.entity.EntityType;
+ import net.minecraft.world.entity.monster.Vindicator;
+@@ -927,6 +928,37 @@ public class PaperWorldConfig {
+         return table;
+     }
+ 
++    public it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<ResourceLocation> featureSeeds = new it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<>();
++    private void featureSeeds() {
++        featureSeeds.defaultReturnValue(-1);
++        boolean randomise = getBoolean("feature-seeds.generate-random-seeds-for-all", false);
```
Comment (Proximyst) [inline patches/server/0802-Configurable-feature-seeds.patch]: Please use `final` on local variables.

Is it intentional for this to not be set to false again after randomising?

**Review by Proximyst — COMMENTED** [review]
looks pretty fine to me

File: patches/server/0802-Configurable-feature-seeds.patch (hunk 1 above)
Comment (kennytv) [reply patches/server/0802-Configurable-feature-seeds.patch]: We'd have to know that this is the last world to be checked to disable this with default values, so it doesn't only generate them for one world. It doesn't have any effect afterwards (unless you go out of your way to delete entries, at which point you might as well set it to false again)

**Comment by MiniDigger** [comment]
https://github.com/PaperMC/PaperDocs/pull/109

File: patches/server/0802-Configurable-feature-seeds.patch [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
++            section = config.getConfigurationSection("world-settings.default.feature-seeds");
++        }
++        for (final String key : section.getKeys(false)) {
++            net.minecraft.resources.ResourceLocation resourceKey = new net.minecraft.resources.ResourceLocation(key);
++            featureSeeds.put(resourceKey, section.getLong(key));
++        }
++        if (randomise) {
++            final Map<String, Object> randomisedSeeds = new HashMap<>();
++            final java.util.Random random = new java.util.Random();
++            for (net.minecraft.resources.ResourceLocation resourceLocation : net.minecraft.server.MinecraftServer.getServer()
```
Comment (Proximyst) [inline patches/server/0802-Configurable-feature-seeds.patch]: final :p

**Review by Proximyst — APPROVED** [review]
knenwald quality 🐒

**Comment by LogGits** [comment]
Heya is it possible to get some type of example of the usage of this, If i understand correctly this allows us to change the seed generation of almost anything? If so how for I go about setting different generation for:
```
- Clay
- Ores
- Spawner Generation
- Cave Generation
- Slime chunks
```
And if i wanted to get the randomly generated seed for those things, is that possible via the api or is it randomised every time server starts up?

**Pull request merged**